src/main/java/com/google/step/snippet/servlets/SearchServlet.java

```diff
@@ -79,13 +84,53 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
     }
     String query = encodeValue(param);
     List<Card> allCards = new ArrayList<>();
-    for (Client client : clients) {
-      String link = getLink(client.getCseId(), query);
-      Card card = client.search(link);
-      if (card != null) {
-        allCards.add(card);
-      }
+    ExecutorService executor = Executors.newFixedThreadPool(3);
+    List<Callable<Card>> cardCallbacks =
+        Arrays.asList(
+            () -> {
+              String w3Link = getLink(w3Client.getCseId(), query);
+              if (w3Link != null) {
+                return w3Client.search(w3Link);
+              }
+              return null;
+            },
+            () -> {
+              String stackLink = getLink(stackClient.getCseId(), query);
+              if (stackLink != null) {
+                return stackClient.search(stackLink);
+              }
+              return null;
+            },
+            () -> {
+              String geeksLink = getLink(geeksClient.getCseId(), query);
+              if (geeksLink != null) {
+                return geeksClient.search(geeksLink);
+              }
+              return null;
+            });
+    /* thread process completion takes roughly 1.5 to 2.5 seconds, thus a timeout of
+     * 3 seconds was chosen.
+     */
+    try {
+      executor.invokeAll(cardCallbacks, 3L, TimeUnit.SECONDS).stream()
+          .map(
+              future -> {
+                try {
+                  return future.get();
+                } catch (CancellationException | ExecutionException | InterruptedException e) {
+                  return null;
```

Comment:
Shouldn't be returning null here either?

Reply:
The code here is performing an operation (future.get()) on each Future in the list of Future's returned by invokeAll(). I'd say returning null is appropriate here if this operation fails on one of the Future's, because we'd want it to continue through the list.